File: 14-2.rb
#!/usr/bin/env ruby

class Polymer
  attr_accessor :pairs
  attr_accessor :rules

  def initialize(template)
    @pairs = Hash.new 0
    @rules = {}
    @trail = template[-1]

    (template.size - 1).times do |pos|
      pair = template[pos, 2]
      @pairs[pair] += 1
    end
  end

  def add_rule(pair, insert)
    @rules[pair] = insert
  end

  def step
    new_pairs = Hash.new 0
    @pairs.each do |pair, count|
      if production = @rules[pair]
        new_pairs[pair[0] + production] += count
        new_pairs[production + pair[1]] += count
      else
        new_pairs[pair] = count
      end
    end
    @pairs = new_pairs
  end

  def score
    counts = Hash.new 0
    @pairs.each { |pair, count| counts[pair[0]] += count }
    counts[@trail] += 1

    l = counts.values.sort
    l[-1] - l[0]
  end

  def size
    @pairs.values.sum + 1
  end

  def to_s
    "<#{self.class}>"
  end

  def inspect
    to_s
  end
end

insertion_format = /^(?<pair>[[:upper:]]{2}) -> (?<insert>[[:upper:]])$/

input = File.read('14.input').lines.map(&:strip)

p = Polymer.new input.shift

input.shift

input.each do |line|
  insertion = insertion_format.match line
  p.add_rule insertion['pair'], insertion['insert']
end

40.times do
  p.step
end

print p.score, "\n"

File: 14-2.py
#!/usr/bin/env python

from collections import deque
import re

class Polymer:
    def __init__(self, template):
        self.pairs = {}
        self.rules = {}
        self.trail = template[-1]

        for pos in range(len(template) - 1):
            pair = template[pos : pos + 2]
            self.pairs[pair] = self.pairs.get(pair, 0) + 1

    def add_rule(self, pair, insert):
        self.rules[pair] = insert

    def step(self):
        new_pairs = {}
        for pair, count in self.pairs.items():
            production = self.rules.get(pair, None)
            if production is not None:
                new_pairs[pair[0] + production] = new_pairs.get(pair[0] + production, 0) + count
                new_pairs[production + pair[1]] = new_pairs.get(production + pair[1], 0) + count
            else:
                new_pairs[pair] = count
        self.pairs = new_pairs

    def score(self):
        counts = {}
        for pair, count in self.pairs.items():
            counts[pair[0]] = counts.get(pair[0], 0) + count
        counts[self.trail] = counts.get(self.trail, 0) + 1
        l = list(counts.values())
        l.sort()
        return l[-1] - l[0]

    def __str__(self):
        return f"<{self.__class__.__name__}: {self.molecule}>"


insertion_format = re.compile("(?P<pair>[A-Z]{2}) -> (?P<insert>[A-Z])")

input = deque([line.strip() for line in open('14.input').readlines()])

p = Polymer(input.popleft())

input.popleft()

for line in input:
    insertion = insertion_format.match(line)
    p.add_rule(insertion['pair'], insertion['insert'])

for _ in range(40):
    p.step()

print(p.score())
